crates/partition-store/src/persisted_lsn_tracking.rs

```diff
@@ -0,0 +1,238 @@
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
+use std::io::Cursor;
+
+use rocksdb::event_listener::{EventListener, FlushJobInfo};
+use rocksdb::table_properties::{
+    EntryType, TablePropertiesCollector, TablePropertiesCollectorFactory,
+};
+use tokio::sync::mpsc;
+use tracing::warn;
+
+use crate::fsm_table::{PartitionStateMachineKey, SequenceNumber, fsm_variable};
+use crate::keys::{KeyKind, TableKey};
+use crate::protobuf_types::PartitionStoreProtobufValue;
+use restate_types::{identifiers::PartitionId, logs::Lsn};
+
+pub(crate) struct LatestAppliedLsnCollector {
+    applied_lsns: HashMap<PartitionId, Lsn>,
+}
+
+impl LatestAppliedLsnCollector {
+    fn new() -> Self {
+        LatestAppliedLsnCollector {
+            applied_lsns: Default::default(),
+        }
+    }
+}
+
+impl TablePropertiesCollector for LatestAppliedLsnCollector {
+    fn add_user_key(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        entry_type: rocksdb::table_properties::EntryType,
+        _seq: u64,
+        _file_size: u64,
+    ) -> bool {
+        if matches!(entry_type, EntryType::EntryPut) {
+            if let Some((partition_id, lsn)) = extract_partition_applied_lsn(key, value) {
+                self.applied_lsns
+                    .entry(partition_id)
+                    .and_modify(|existing| {
+                        if lsn > *existing {
+                            *existing = lsn;
+                        }
+                    })
+                    .or_insert(lsn);
+            }
+        }
+        true
+    }
+
+    fn finish(&mut self, properties: &mut HashMap<CString, CString>) -> bool {
+        for (partition_id, lsn) in &self.applied_lsns {
+            properties.insert(
+                CString::new(applied_lsn_property_name(*partition_id)).unwrap(),
+                CString::new(lsn.as_u64().to_string()).unwrap(),
+            );
+        }
+        true
+    }
+
+    fn get_readable_properties(&self) -> HashMap<CString, CString> {
+        let mut properties = HashMap::new();
+        for (partition_id, lsn) in &self.applied_lsns {
+            properties.insert(
+                CString::new(applied_lsn_property_name(*partition_id)).unwrap(),
+                CString::new(lsn.as_u64().to_string()).unwrap(),
+            );
+        }
+        properties
+    }
```

Comment:
We probably don't care that much about this; it's only for logging, and only if RocksDB's log level is debug.

Reply:
How often will this be called? Looks like it allocates every time a new HashMap. But I think it is ok to take some performance hit if people activate the debug log level explicitly.